#### 异常

- Rust中的异常有4中：Option、Result、Panic、Abort
- Option用于应对可能失败的情况，用Some和None来表示是否失败。
- Result用于应对可恢复错误，Rust用成功或失败来表示是否有错误。出错不一定导致程序崩溃，但需要进行专门处理，以使程序继续执行。
- Panic是Rust提供的一种机制，用于在遇到不可恢复的错误时进行清理内存。
- 当遇到Panic那样的错误时，如果不想用Panic，而是想让操作系统来清理内存，则可以使用Abort。

#### 异常处理

- 如果遇到错误时只想向上抛，不想处理，则可以使用"?". 此时返回类型是Result，成功时返回String，错误时返回io:Error。
- 其实Option和Result非常相似，Option<T>可以看成Result<T,()>

``` rust
use std::fs::File;
use std::io;
use std::io::{Error, Read, read_to_string};

fn main() {
    let s = read_from_file();
    match s {
        Ok(e) => println!("{}", e),
        Ok(s) => println!("{}", s),
        _ => {}
    }
}

fn read_from_file() -> Result<String, io::Error> {
    let mut f = File::open("./key.txt")?; // 出错时直接抛出
    let mut s = String::new();
    f.read_to_string(&mut s).expect("TODO: panic message");
    Ok(s)
}
```